lib/authenticators/okta/impl.go

```diff
@@ -0,0 +1,261 @@
+	factorsVerifyPathExtra = "/factors/%s/verify"
+)
+
+type VerifyTOTPFactorDataType struct {
+	StateToken string `json:"stateToken,omitempty"`
+	PassCode   string `json:"passCode,omitempty"`
+}
+
+type LoginDataType struct {
+	Password string `json:"password,omitempty"`
+	Username string `json:"username,omitempty"`
+}
+
+type MFAFactorsType struct {
+	Id         string `json:"id,omitempty"`
+	FactorType string `json:"factorType,omitempty"`
+	Provider   string `json:"provider,omitempty"`
+	VendorName string `json:"vendorName,omitempty"`
+}
+
+type UserProfileType struct {
+	Login string `json:"login,omitempty"`
+}
+
+type UserInfoType struct {
+	Id      string          `json:"id,omitempty"`
+	Profile UserProfileType `json:"profile,omitempty"`
+}
+
+type EmbeddedDataResponseType struct {
+	User   UserInfoType     `json:"user,omitempty"`
+	Factor []MFAFactorsType `json:"factors,omitempty"`
+}
+
+type PrimaryResponseType struct {
+	StateToken      string                   `json:"stateToken,omitempty"`
+	ExpiresAtString string                   `json:"expiresAt,omitempty"`
+	Status          string                   `json:"status,omitempty"`
+	Embedded        EmbeddedDataResponseType `json:"_embedded,omitempty"`
+}
+
+type PushResponseType struct {
+	ExpiresAtString string                   `json:"expiresAt,omitempty"`
+	Status          string                   `json:"status,omitempty"`
+	FactorResult    string                   `json:"factorResult,omitempty"`
+	Embedded        EmbeddedDataResponseType `json:"_embedded,omitempty"`
+}
+
+func newPublicAuthenticator(oktaDomain string, logger log.Logger) (
+	*PasswordAuthenticator, error) {
+	return &PasswordAuthenticator{
+		authnURL:   fmt.Sprintf(authEndpointFormat, oktaDomain),
+		logger:     logger,
+		recentAuth: make(map[string]authCacheData),
+	}, nil
+}
+
+func (pa *PasswordAuthenticator) passwordAuthenticate(username string,
+	password []byte) (bool, error) {
+	loginData := LoginDataType{Password: string(password), Username: username}
+	body := &bytes.Buffer{}
+	encoder := json.NewEncoder(body)
+	encoder.SetIndent("", "    ") // Make life easier for debugging.
+	if err := encoder.Encode(loginData); err != nil {
+		return false, err
+	}
+	req, err := http.NewRequest("POST", pa.authnURL, body)
+	if err != nil {
+		return false, err
+	}
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Content-Type", "application/json")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode == http.StatusUnauthorized {
+		return false, nil
+	}
+	if resp.StatusCode != http.StatusOK {
+		return false, fmt.Errorf("bad status: %s", resp.Status)
+	}
+	decoder := json.NewDecoder(resp.Body)
+	var response PrimaryResponseType
+	if err := decoder.Decode(&response); err != nil {
+		return false, err
+	}
+	// TODO: change logger to debug capable
+	//pa.logger.Printf("oktaresponse=%+v", response)
```

Comment:
Please change the API to require a `DebugLogger`. The caller already passes one in.

Reply:
Yes, please change the API.